draco/asp/hard.lp

```diff
@@ -192,4 +191,48 @@ violation(aggregate_not_all_continuous):-
     E1 > E2,
     attribute((encoding,scale_type),E2,T),
     domain(continuous_scale,T),
+    not attribute((encoding,binning),E2,_),
     not attribute((encoding,aggregate),E2,_).
+
+% @hard(detail_not_ordinal) Detail requires ordinal scales.
+violation(detail_not_ordinal) :-
+    attribute((scale,channel),S,detail),
+    not attribute((scale,type),S,ordinal).
+
+% @hard(bar_tick_area_line_without_continuous_x_y) Bar, tick, line, area require some continuous variable on x or y.
+violation(bar_tick_area_line_without_continuous_x_y) :-
+    attribute((mark,type),M,(bar;tick;area;line)),
+    attribute(mark_channel_discrete_or_binned,M,x),
+    attribute(mark_channel_discrete_or_binned,M,y).
+
+% @hard(zero_d_n) Can only use zero with datetime or number.
+violation(zero_d_n) :-
+    attribute((mark,scale),M,S),
+    entity(encoding,M,E),
+    attribute((scale,zero),S,_),
+    attribute((encoding,field),E,F),
+    attribute((field,type),F,T),
+    T != datetime,
+    T != number.
+
+% @hard(bar_area_without_zero) Bar and area mark requires scale of continuous to start at zero.
+violation(bar_area_without_zero) :-
+    attribute((mark,type),M,(bar;area)),
+    entity(encoding,M,E),
+    attribute((encoding,channel),E,x),
+    not attribute((encoding,binning),E,_),
+    attribute((mark,scale),M,S),
+    attribute((scale,channel),S,x),
```

Comment:
can we write `x;y`?

Reply:
I don't think so, the channel need to be exact the same across encoding and scale

Reply:
Oh I missed that. Could you use a variable and then say that the variable is x;y?